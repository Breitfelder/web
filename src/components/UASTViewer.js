import React, { Component } from 'react';
import FlatUASTViewer from 'uast-viewer';

const ROOT_ID = 1;

export const getSearchResults = uast => {
  if (!uast) {
    return null;
  }

  const rootNode = uast[ROOT_ID];
  if (!rootNode) {
    return null;
  }

  if (Array.isArray(rootNode.n)) {
    return rootNode.n.map(c => c.id);
  }

  return null;
};

class UASTViewer extends Component {
  render() {
    const { uastViewerProps, showLocations } = this.props;
    const { flatUast } = uastViewerProps;

    if (!flatUast) {
      return null;
    }

    const searchResults = getSearchResults(flatUast);
    let rootIds = searchResults || [ROOT_ID];

    if (searchResults && !searchResults.length) {
      return <div className="bblfsh-uast-viewer__not-found">Nothing found</div>;
    }

    return (
      <FlatUASTViewer
        {...uastViewerProps}
        rootIds={rootIds}
        showLocations={showLocations}
        style={{ overflow: 'auto' }}
      />
    );
  }
}

export default UASTViewer;
