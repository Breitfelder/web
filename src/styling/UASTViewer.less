.bblfsh-uast-viewer {
    &__not-found {
        width: 100%;
        padding: 1rem 0;
        text-align: center;
    }
}
