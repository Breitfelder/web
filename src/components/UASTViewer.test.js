import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import UASTViewer, { SEARCH_RESULTS_TYPE } from './UASTViewer';

const shouldMatchSnapshot = comp => {
  const wrapper = renderer.create(comp);
  expect(wrapper.toJSON()).toMatchSnapshot();
};

it('empty render', () => {
  shouldMatchSnapshot(<UASTViewer uastViewerProps={{ uast: {} }} />);
});

it('tree render', () => {
  const uastViewerProps = {
    uast: {
      1: { id: 1, expanded: true, StartPosition: { Line: 1, Col: 2 } },
    },
  };

  shouldMatchSnapshot(
    <UASTViewer uastViewerProps={uastViewerProps} showLocations={false} />
  );
});

it('tree render with locations', () => {
  const uastViewerProps = {
    uast: {
      1: { id: 1, expanded: true, StartPosition: { Line: 1, Col: 2 } },
    },
  };

  shouldMatchSnapshot(
    <UASTViewer uastViewerProps={uastViewerProps} showLocations={true} />
  );
});

describe('search results', () => {
  it('render', () => {
    const uastViewerProps = {
      uast: {
        1: { id: 1, InternalType: SEARCH_RESULTS_TYPE, Children: [2] },
        2: { id: 2, InternalType: 'TestType' },
      },
    };

    shouldMatchSnapshot(<UASTViewer uastViewerProps={uastViewerProps} />);
  });

  it('empty', () => {
    const uastViewerProps = {
      uast: {
        1: { id: 1, InternalType: SEARCH_RESULTS_TYPE, Children: [] },
      },
    };

    shouldMatchSnapshot(<UASTViewer uastViewerProps={uastViewerProps} />);
  });
});
