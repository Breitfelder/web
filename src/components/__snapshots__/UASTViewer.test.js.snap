// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty render 1`] = `
<div
  className="uast-viewer"
  onMouseLeave={[Function]}
  style={
    Object {
      "overflow": "auto",
    }
  }
/>
`;

exports[`search results empty 1`] = `
.c0 {
  width: 100%;
  padding: 1rem 0;
  text-align: center;
}

<div
  className="c0"
>
  Nothing found
</div>
`;

exports[`search results render 1`] = `
<div
  className="uast-viewer"
  onMouseLeave={[Function]}
  style={
    Object {
      "overflow": "auto",
    }
  }
>
  <div
    className="uast-viewer__node-2 uast-item"
    onClick={[Function]}
    onMouseMove={[Function]}
  >
    <div
      className="uast-title uast-title-collapsible collapsed"
      onClick={[Function]}
    >
      <summary
        className="uast-label"
      >
        Node
      </summary>
    </div>
  </div>
</div>
`;

exports[`tree render 1`] = `
<div
  className="uast-viewer"
  onMouseLeave={[Function]}
  style={
    Object {
      "overflow": "auto",
    }
  }
>
  <div
    className="uast-viewer__node-1 uast-item"
    onClick={[Function]}
    onMouseMove={[Function]}
  >
    <div
      className="uast-title uast-title-collapsible "
      onClick={[Function]}
    >
      <summary
        className="uast-label"
      >
        Node
      </summary>
    </div>
    <div
      className="uast-content"
    />
  </div>
</div>
`;

exports[`tree render with locations 1`] = `
<div
  className="uast-viewer"
  onMouseLeave={[Function]}
  style={
    Object {
      "overflow": "auto",
    }
  }
>
  <div
    className="uast-viewer__node-1 uast-item"
    onClick={[Function]}
    onMouseMove={[Function]}
  >
    <div
      className="uast-title uast-title-collapsible "
      onClick={[Function]}
    >
      <summary
        className="uast-label"
      >
        Node
      </summary>
    </div>
    <div
      className="uast-content"
    >
      <div
        className="undefined uast-item"
      >
        <div
          className="uast-title uast-title-collapsible "
          onClick={[Function]}
        >
          <span
            className="uast-property-name"
          >
            start_position
          </span>
          <summary
            className="uast-label"
          >
            Position
          </summary>
        </div>
        <div
          className="uast-content"
        >
          <div
            className="undefined uast-item"
          >
            <div
              className="uast-title"
            >
              <span
                className="uast-property-name"
              >
                line
              </span>
              <span
                className="uast-value number"
              >
                1
              </span>
            </div>
          </div>
          <div
            className="undefined uast-item"
          >
            <div
              className="uast-title"
            >
              <span
                className="uast-property-name"
              >
                col
              </span>
              <span
                className="uast-value number"
              >
                2
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
